src/Processors/QueryPlan/PartsSplitter.cpp

```diff
@@ -216,18 +284,82 @@ struct PartsRangesIterator
         return true;
     }
 
+    void dump(WriteBuffer & buffer) const
+    {
+        buffer << "Part index " << part_index;
+        buffer << " event " << (event == PartsRangesIterator::EventType::RangeStart ? "Range Start" : "Range End");
+        buffer << " range begin " << range.begin;
+        buffer << " end " << range.end;
+        buffer << " value " << ::toString(value) << '\n';
+    }
+
+    [[maybe_unused]] String toString() const
+    {
+        WriteBufferFromOwnString buffer;
+        dump(buffer);
+        return buffer.str();
+    }
+
     Values value;
     MarkRange range;
     size_t part_index;
     EventType event;
 };
 
+struct PartRangeIndex
+{
+    explicit PartRangeIndex(PartsRangesIterator & ranges_iterator)
+        : part_index(ranges_iterator.part_index)
+        , range(ranges_iterator.range)
+    {}
+
+    bool operator==(const PartRangeIndex & other) const
+    {
+        return part_index == other.part_index && range.begin == other.range.begin && range.end == other.range.end;
+    }
+
+    bool operator<(const PartRangeIndex & other) const
+    {
+        return part_index < other.part_index && range.begin < other.range.begin && range.end < other.range.end;
```

Comment:
why it is not equivalent to `tuple(part_index, begin, end) < tuple(other...)`? pls leave a comment.

Reply:
Updated.